### PR TITLE
Custom theme

### PR DESCRIPTION
**Pull request by iTTou**
Let the user change the default theme primary color, switch between rounded and squared edges and swap the default dark theme with a light one

### CODE REVIEW

File: WoWsShipBuilder.Common/Infrastructure/ApplicationTheme/ThemeManager.cs [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+
+    public MudTheme AppTheme { get; private set; }
+
+    private MudTheme GetAppTheme()
+    {
+        this.SelectedThemeVariant = Helpers.IsAprilFool() ? ThemeVariant.AprilFool : this.appSettings.ThemeVariant ?? DefaultThemeVariant;
+
+        return this.SelectedThemeVariant switch
+        {
+            ThemeVariant.Dark => this.CreateDarkTheme(),
```
Comment (floribe2000) [inline WoWsShipBuilder.Common/Infrastructure/ApplicationTheme/ThemeManager.cs]: There is no need to create two themes for light and dark. Simply use the Palette and PaletteDark properties on the MudTheme to set the two themes.
April fools theme can stay a separate theme

File: WoWsShipBuilder.Common/Infrastructure/ApplicationTheme/ThemeManager.cs [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+    public ThemeManager(AppSettings appSettings)
+    {
+        this.appSettings = appSettings;
+        this.AppTheme = this.GetAppTheme();
+    }
+
+    public event EventHandler<MudTheme>? ThemeChanged;
+
+    public enum ThemeVariant
```
Comment (floribe2000) [inline WoWsShipBuilder.Common/Infrastructure/ApplicationTheme/ThemeManager.cs]: move public enums to their own file

File: WoWsShipBuilder.Common/Features/ShipSelection/ShipSelector.razor [hunk 3]
```diff
@@ -136,7 +136,7 @@
                             <MudListItem key="@shipContainer.Id">
                                 <MudStack Style="align-items: stretch" Spacing="1" Class="my-n1">
                                     <MudStack Row Justify="Justify.SpaceBetween">
-                                        <MudText>
+                                        <MudText Color="Color.Primary">
```
Comment (floribe2000) [inline WoWsShipBuilder.Common/Features/ShipSelection/ShipSelector.razor:139]: Why this change? I don't think changing the color to the primary here is a good choice

File: WoWsShipBuilder.Web/Features/Layout/TopMenu.razor [hunk 4]
```diff
@@ -206,6 +204,9 @@
 
 @code
 {
+    [Parameter, EditorRequired]
+    public bool IsDarkTheme { get; set; }
```
Comment (floribe2000) [inline WoWsShipBuilder.Web/Features/Layout/TopMenu.razor]: This is styling information that should not have to be passed to the component

**Review by floribe2000 — CHANGES_REQUESTED** [review]
Overall, this looks good.
However, theme information should not be passed to all the individual components. A component should be independent from the selected theme. This means that ThemeManager should only be available to the top level component where the theme is actually applied but should not be passed to ANY other component or page.
In cases where it is absolutely necessary (like applying the filter to icons/images) consider creating a new component where the ThemeManager is injected. This reduces coupling between theming and the rest of the app

File: WoWsShipBuilder.Common/Features/ShipSelection/ShipSelector.razor (hunk 3 above)
Comment (iTTou) [reply WoWsShipBuilder.Common/Features/ShipSelection/ShipSelector.razor:139]: It helps reading the list when serveral ships are selected . in particular when the light theme is selected and / or there are builds applied

File: WoWsShipBuilder.Common/Features/Builds/Components/BuildImageRenderDialog.razor [hunk 5]
```diff
@@ -4,6 +4,7 @@
 @using WoWsShipBuilder.Features.ShipStats.ViewModels
 @using WoWsShipBuilder.Features.Settings
 @using WoWsShipBuilder.Infrastructure.ApplicationData
+@using WoWsShipBuilder.Infrastructure.ApplicationTheme
```
Comment (floribe2000) [inline WoWsShipBuilder.Common/Features/Builds/Components/BuildImageRenderDialog.razor]: remove unused using directive

File: WoWsShipBuilder.Common/Features/ShipStats/Components/ShipModuleSelector.razor [hunk 6]
```diff
@@ -11,12 +13,13 @@
...
                     @foreach (var upgrade in columnUpgrade)
                     {
+                        var isSelected = ViewModel.SelectedModules.Contains(upgrade);
                         if (columnUpgrade.IndexOf(upgrade) != 0)
                         {
                             <MudIcon Icon="@Icons.Material.Filled.ArrowDropDown" Class="mt-n3 mb-n2"/>
                         }
                         <MudButton OnClick="@(_ => ViewModel.SelectModuleExecute(upgrade))" Class="pa-2 mb-n2" Style="@GetImageStyle()" Variant="Variant.Outlined" >
-                            <MudImage Width="@GetImageSize()" Height="@GetImageSize()" Src="@GetModuleIcon(upgrade, ViewModel.SelectedModules.Contains(upgrade))"/>
+                            <MudImage Width="@GetImageSize()" Height="@GetImageSize()" Src="@GetModuleIcon(upgrade, isSelected)" Class="image-filter-invert-50"/>
```
Comment (floribe2000) [inline WoWsShipBuilder.Common/Features/ShipStats/Components/ShipModuleSelector.razor]: applying the image filter here results in the golden icons (for selected modules) being inverted to gray when in light theme.
In this case, it is probably necessary to set the class depending on the selection state of the element, similar to how the Src is set

File: WoWsShipBuilder.Common/Features/ShipStats/Components/ShipStatsTab.razor [hunk 7]
```diff
@@ -37,6 +35,7 @@
 @inject IUserDataService UserDataService
 @inject ISnackbar Snackbar
 @inject AppNavigator AppNavigator
+@inject ThemeManager ThemeManager
```
Comment (floribe2000) [inline WoWsShipBuilder.Common/Features/ShipStats/Components/ShipStatsTab.razor]: remove unused injection

File: WoWsShipBuilder.Common/Features/ShipStats/Components/ShipModuleSelector.razor [hunk 8]
```diff
@@ -1,6 +1,8 @@
 ﻿@using WoWsShipBuilder.DataStructures.Ship
+@using WoWsShipBuilder.Infrastructure.ApplicationTheme
 @inherits ReactiveComponentBase<WoWsShipBuilder.Features.ShipStats.ViewModels.ShipModuleViewModel>
 @inject ILocalizer Localizer
+@inject ThemeManager ThemeManager
```
Comment (floribe2000) [inline WoWsShipBuilder.Common/Features/ShipStats/Components/ShipModuleSelector.razor]: remove unused injection

File: WoWsShipBuilder.Desktop/Features/BlazorWebView/DesktopLayout.razor [hunk 9]
```diff
@@ -1,7 +1,13 @@
...
 @inherits LayoutComponentBase
+@implements IDisposable
+
+@inject ThemeManager ThemeManager
+@inject AppSettings AppSettings
 
-<MudThemeProvider Theme="@theme" IsDarkMode="true"/>
+<MudThemeProvider Theme="@theme" IsDarkMode="isDarkTheme"/>
 <MudDialogProvider/>
 <MudSnackbarProvider/>
```
Comment (floribe2000) [inline WoWsShipBuilder.Desktop/Features/BlazorWebView/DesktopLayout.razor:14]: The custom theme provider component is missing for desktop

File: WoWsShipBuilder.Desktop/Features/BlazorWebView/DesktopLayout.razor [hunk 10]
```diff
@@ -13,13 +19,41 @@
         </CascadingValue>
     </MudMainContent>
     <footer style="margin-top: 16px;">
-        <ShipBuilderFooter/>
+        <ShipBuilderFooter IsDarkTheme="isDarkTheme"/>
```
Comment (floribe2000) [inline WoWsShipBuilder.Desktop/Features/BlazorWebView/DesktopLayout.razor]: this is ok, a nicer solution would have been to also solve this entirely through css. But since this component is only ever used from the main layout, it's fine

File: WoWsShipBuilder.Desktop/Features/Settings/SettingsComponent.razor [hunk 11]
```diff
@@ -82,38 +85,55 @@
...
+                <MudSelect T="ThemeEnums.ThemeStyle" @bind-Value="themeStyle" Dense Immediate Label="@Localizer.SimpleAppLocalization(nameof(Translation.Settings_ThemeStyle))" Disabled="@Helpers.IsAprilFool()">
+                    @foreach (var style in Enum.GetValues<ThemeEnums.ThemeStyle>())
+                    {
+                        <MudSelectItem Value="@style">
+                            <MudText Color="Color.Default">@Localizer.GetAppLocalization(style.ToString()).Localization</MudText>
+                        </MudSelectItem>
+                    }
+                </MudSelect>
+                <div class="mt-2" style="@($"background-color:{themePrimaryColor}; height: 36.5px; width:36.5px; border-radius: var(--mud-default-borderradius);")"></div>
+                <MudColorPicker Label="@Localizer.SimpleAppLocalization(nameof(Translation.Settings_ThemePrimaryColor))" DisableToolbar="false" PickerVariant="PickerVariant.Dialog" @bind-Value="themePrimaryColor" Style="@($"color: {themePrimaryColor};")" Disabled="@Helpers.IsAprilFool()"/>
```
Comment (floribe2000) [inline WoWsShipBuilder.Desktop/Features/Settings/SettingsComponent.razor:104]: maybe add a button to reset the custom primary color to the theme default?

File: WoWsShipBuilder.Web/Features/UserSettings/UserSettings.razor [hunk 12]
```diff
@@ -52,6 +55,28 @@
...
+                <MudSelect T="ThemeEnums.ThemeStyle" @bind-Value="themeStyle" Dense Immediate Label="@Localizer.SimpleAppLocalization(nameof(Translation.Settings_ThemeStyle))" Disabled="@Helpers.IsAprilFool()">
+                    @foreach (var style in Enum.GetValues<ThemeEnums.ThemeStyle>())
+                    {
+                        <MudSelectItem Value="@style">
+                            <MudText Color="Color.Default">@Localizer.GetAppLocalization(style.ToString()).Localization</MudText>
+                        </MudSelectItem>
+                    }
+                </MudSelect>
+                <div class="mt-2" style="@($"background-color:{themePrimaryColor}; height: 36.5px; width:36.5px; border-radius: var(--mud-default-borderradius);")"></div>
+                <MudColorPicker Label="@Localizer.SimpleAppLocalization(nameof(Translation.Settings_ThemePrimaryColor))" DisableToolbar="false" PickerVariant="PickerVariant.Dialog" @bind-Value="themePrimaryColor" Style="@($"color: {themePrimaryColor};")" Disabled="@Helpers.IsAprilFool()"/>
```
Comment (floribe2000) [inline WoWsShipBuilder.Web/Features/UserSettings/UserSettings.razor:78]: same as for desktop, consider adding a button to reset the custom primary color

**Review by floribe2000 — CHANGES_REQUESTED** [review]
looks good, there are some minor things left but the rest is really nice

**Review by floribe2000 — APPROVED** [review]
Looks good but you missed one of the comments from the previous review. Once that is fixed, feel free to merge
